Test lib for the injection mechanism.

danielfenner opened the pull request:
In order to load shared object into attached process we need a simple
.so first to test / develop that. Thats why the produced binary is
checked in to the testdata folder.

bug: http://b/181304023
test: NA

dpallotti commented on src/UserSpaceInstrumentation/CMakeLists.txt:
```diff
@@ -33,6 +33,22 @@ target_link_libraries(UserSpaceInstrumentation PUBLIC
         OrbitBase
         CONAN_PKG::abseil)
 
+# Test Lib is merely used in UserSpaceInstrumentationTests below. A binary
+# libTestLib.so created from this target is checked in to testdata folder and
+# that binary is used to test the injection mechanism.
+add_library(TestLib SHARED)
```
The name of the library should be expanded, for example to `UserSpaceInstrumentationTestLib`. If I'm not mistaken this still ends up in the `lib` folder and `libTestLib.so` is too generic.

danielfenner replied:
Yep, that's true. UserSpaceInstrumentationTestLib it is.

dpallotti commented on src/UserSpaceInstrumentation/TestLib.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void UseTestLib(const std::string& s) {
+  auto p = GetTmpFilePath();
+  std::ofstream ofs;
```
I get that this is just for testing, but still I thought we wanted to get rid of `std` streams because of exceptions? I guess we should be consistent. There are utilities in `OrbitBase` to read and write an entire text file (after all they are nicer to use than stream directly).

danielfenner replied:
I'd like to keep this as "vanilla" as possible for so I decided against using OrbitBase. 

This will not be linked to Orbit or even any test so I think exceptions don't matter in this case.

dpallotti commented on src/UserSpaceInstrumentation/TestLib.h:
```diff
@@ -0,0 +1,28 @@
+namespace orbit_user_space_instrumentation {
+
+// Library functions for some trivial logging. This library is merely used in tests: A binary
+// produced by this code is checked in into the testdata folder. The test injects this binary into
```
Do you actually need to put it into the testdata folder? The library ends up in the `lib` directory anyway and you can reference it. This is what we did with `LinuxTracingIntegrationTestPuppetSharedObject` and it works fine. It also makes sure that the .so and the code are actually always in sync.

danielfenner replied:
Thanks! Indeed this looks way better.

Adjusted comments and removed the binary in testdata .

dpallotti commented on src/UserSpaceInstrumentation/TestLib.cpp:
```diff
@@ -0,0 +1,52 @@
+  std::cout << "Init Lib. Tmp file is: " << p << std::endl;
+}
+
+void UseTestLib(const std::string& s) {
```
Not that it matters, but I guess you can use `string_view` in this case.

danielfenner replied:
done

danielfenner commented:
Besides addressing the comments I also removed the namespace in the lib; I'd like to have this as easy to work with as possible (nice short symbol names).

dpallotti commented on src/UserSpaceInstrumentation/CMakeLists.txt:
```diff
@@ -33,6 +33,22 @@ target_link_libraries(UserSpaceInstrumentation PUBLIC
         OrbitBase
         CONAN_PKG::abseil)
 
+# Test Lib is merely used in UserSpaceInstrumentationTests below. The binary
```
Change "Test Lib" to the name of the CMake library or to "This test lib".

danielfenner replied:
done

dpallotti commented on src/UserSpaceInstrumentation/UserSpaceInstrumentationTestLib.h:
```diff
@@ -0,0 +1,23 @@
+#define USER_SPACE_INSTRUMENTATION_TEST_LIB_H_
+
+#include <string_view>
+
```
> I also removed the namespace in the lib; I'd like to have this as easy to work with as possible (nice short symbol names).

Maybe add a comment around here with this explanation so that the reason why the namespace is missing doesn't get lost.

danielfenner replied:
As mentioned below: This will change quite a bit no matter what and it really doesn't matter what namespace this is in since it will not link to anything.

dpallotti commented on src/UserSpaceInstrumentation/UserSpaceInstrumentationTestLib.h:
```diff
@@ -0,0 +1,23 @@
+// Log a string into a temporary file.
+void UseTestLib(std::string_view s);
+
+// Call to end using the library. Prints the entire log to standard out and removes the temporary
```
out -> output

danielfenner replied:
done

dpallotti reviewed (approved):
> I'd like to keep this as "vanilla" as possible for so I decided against using OrbitBase.
>
> This will not be linked to Orbit or even any test so I think exceptions don't matter in this case.

Is there a specific gain you have in mind with keeping it vanilla? My personal feeling is that if we are afraid of linking OrbitBase we might be doing something wrong in OrbitBase, and that consistency (in this case, not using streams) is helpful in general, even if this is just a test. (But in any case I don't feel strongly about it.)

danielfenner commented:
> Is there a specific gain you have in mind with keeping it vanilla? My personal feeling is that if we are afraid of linking OrbitBase we might be doing something wrong in OrbitBase, and that consistency (in this case, not using streams) is helpful in general, even if this is just a test. (But in any case I don't feel strongly about it.)

I really don't know yet. I just want to keep it as simple as possible for experimenting. Good chance I'll make it more similar to a lib we want really want to inject in the next PR.
